rust/frontend/src/function.rs

```diff
@@ -217,63 +189,29 @@ impl<'a, 'm> Builder<'a, 'm> {
 
     /// Calls the function that created from `Builder`.
     pub fn invoke(&mut self) -> Result<TVMRetValue> {
-        self.clone()(())
-    }
-}
-
-impl<'a, 'm> FnOnce<((),)> for Builder<'a, 'm> {
-    type Output = Result<TVMRetValue>;
-    extern "rust-call" fn call_once(self, _: ((),)) -> Self::Output {
         if self.func.is_none() {
             bail!("{}", ErrorKind::FunctionNotFound);
         }
 
-        let mut ret_val = unsafe { mem::uninitialized::<ts::TVMValue>() };
-        let mut ret_type_code = 0 as c_int;
-        if self.arg_buf.is_some() {
-            let arg_buf = self.arg_buf?;
-            let mut num_args = arg_buf.len();
-            let mut values = arg_buf
-                .iter()
-                .map(|tav| tav.value.inner)
-                .collect::<Vec<ts::TVMValue>>();
-            let mut tcodes = arg_buf
-                .iter()
-                .map(|tav| tav.type_code as c_int)
-                .collect::<Vec<_>>();
-
-            if self.ret_buf.is_some() {
-                num_args = num_args + 1;
-                let ret_buf = self.ret_buf?;
-                let (ret_val, ret_type_code) = TVMRetValue::into_tvm_value(ret_buf);
-                values.append(&mut vec![ret_val.inner]);
-                tcodes.append(&mut vec![ret_type_code as c_int]);
-            }
-
-            values.truncate(num_args);
-            tcodes.truncate(num_args);
-            check_call!(ts::TVMFuncCall(
-                self.func?.handle,
-                values.as_mut_ptr(),
-                tcodes.as_mut_ptr(),
-                num_args as c_int,
-                &mut ret_val as *mut _,
-                &mut ret_type_code as *mut _
-            ));
-        } else {
-            check_call!(ts::TVMFuncCall(
-                self.func?.handle,
-                ptr::null_mut(),
-                ptr::null_mut(),
-                0 as c_int,
-                &mut ret_val as *mut _,
-                &mut ret_type_code as *mut _
-            ));
-        }
+        let num_args = self.arg_buf.len();
+        let (mut values, mut type_codes): (Vec<ffi::TVMValue>, Vec<ffi::TVMTypeCode>) = self
+            .arg_buf
+            .iter()
+            .map(|tvm_arg| (tvm_arg.value, tvm_arg.type_code as ffi::TVMTypeCode))
+            .unzip();
+
+        let mut ret_val = TVMValue { v_int64: 4 };
+        let mut ret_type_code = 0;
+        check_call!(ffi::TVMFuncCall(
+            self.func?.handle,
+            values.as_mut_ptr(),
```

Comment:
I think that it's more helpful to let any errors reveal themselves instead of truncate and fix the symptoms but not the underlying cause.

Reply:
No actually it's necessary, because it can cause UB accessing uninitialized memory. Either use truncate by length or shrink_to_fit.

Reply:
> No actually it's necessary, because it can cause UB accessing uninitialized memory.

Nope, `let num_args = self.arg_buf.len()` is the actual (initialized) length of the vec (unless one uses [`set_len`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len) or the like, but that's just asking for trouble). This value is passed  [into `TVMFuncCall`](https://github.com/dmlc/tvm/blob/master/src/runtime/c_runtime_api.cc#L406) where it becomes [part of a `TVMValue`](https://github.com/dmlc/tvm/blob/master/include/tvm/runtime/packed_func.h#L317). If the func is in a TVM module, it'll raise an error that the wrong number of arguments are passed and the user can fix the error instead of being left with a subtly buggy program.

Reply:
I don't understand your resistance towards making passing the pointer safer! The internal RawVec of Vec has capacity and reallocates with uninitialized memory when length matches the capacity. If you don't use `shrink_to_fit` or `truncate` by length, you're potentially allowing UB despite could be unlikely and adding one call to `shrink_to_fit` would remove any possibilities like that.

Reply:
> passing the pointer safer

Hiding errors from the user is not safer, is what I'm trying to say.

> reallocates with uninitialized memory when length matches the capacity

You know that reallocation is impossible during the scope of the mutable borrow, right?

Reply:
Let me explain it again. The `values` is a `Vec`. `Vec` has capacity which is equal to the number of allocated memory and length which is the number of elements actually pushed/inserted. What I'm saying is that you're passing `values.as_mut_ptr()` over FFI, which is pointing to a internal memory buffer `RawVec` with some uninitialized memory. If for some reason the backend accesses an uninitialized memory it'll be UB. Calling `shrink_to_fit` on `values` before `value.as_mut_ptr()` would guarantee that UB cannot happen because of that. Make sense?

Reply:
> Make sense?

No, because the `num_args` (i.e. `values.len()`) is what prevents accessing out-of-bounds elements. There will almost always be memory after the elements of the vec, whether or not they're part of the vec's allocation. A buggy client is always welcome to be buggy; you can't prevent that by truncating the vec.